### PR TITLE
Fix form in session

### LINKED ISSUE #1597: Setting form data in session breaks coupon
Hi,

I'm working currently in Thelia beta 2.2.

When i render a coupon update template (/admin/coupon/update/{$id}) my php session is broken because :

In the [ParserContext::addForm() method](https://github.com/thelia/thelia/blob/master/core/lib/Thelia/Core/Template/ParserContext.php#L124), when adding form data in session there is a problem because the Coupon condition classes all extends [ConditionAbstract](https://github.com/thelia/thelia/blob/master/core/lib/Thelia/Condition/Implementation/ConditionAbstract.php#L42) which contains [BaseFacade](https://github.com/thelia/thelia/blob/master/core/lib/Thelia/Coupon/BaseFacade.php#L40) which contains application container (this object is unserializable and can not be serialized) !

### PR DESCRIPTION
**Pull request by gillesbourgeat**
This pull request offers a solution to avoid passing all forms in the session
The forms is added to the session only if the error url is present

resolve #1597

### CODE REVIEW

**Comment by Alban-io** [comment]
:+1:

File: core/lib/Thelia/Core/Template/ParserContext.php [hunk 1]
```diff
@@ -108,29 +119,31 @@ public function getCurrentForm()
...
-        $formErrorInformation = $this->request->getSession()->getFormErrorInformation();
-
         $this->set(get_class($form) . ":" . $form->getType(), $form);
 
-        /** Remove unserializable objects \Symfony\Component\HttpFoundation\File\UploadedFile */
-        $formData = $form->getForm()->getData();
-        foreach ($formData as $idx => $value) {
-            if ($value instanceof UploadedFile) {
-                unset($formData[$idx]);
+        if ($form->getForm()->get("error_url")->getData() !== null) {
```
Comment (roadster31) [inline core/lib/Thelia/Core/Template/ParserContext.php]: Why don't you use BaseForm::getErrorUrl() ?

**Comment by roadster31** [comment]
This PR is just a workaround, and is not addressing our main concern about forms : defining a standard way to process form errors (specially in front office), and designing a clean and robust form lifecycle, which includes form errors feedback to the user.

**Comment by lunika** [comment]
:+1: with roadster

**Comment by roadster31** [comment]
Additionnaly, the PR is breaking compatibility with situations where error_url is not defined, while the form stored in session is nevertheless used by the   displayed template

**Comment by gillesbourgeat** [comment]
I know that this pr is just a workaround. But it is not possible to pass automatically all forms in the session, for use of this behavior very seldom.
I'm sorry, but I not see a use case for this behavior.
However, it's possible to add a parameter "pushInSeesion" on the method ```addForm```
```php
// if $pushInSeesion is true or error_url is not empty, then add the form in session.
public function addForm(BaseForm $form, $pushInSeesion = false)
```

File: core/lib/Thelia/Core/Template/ParserContext.php (hunk 1 above)
Comment (gillesbourgeat) [inline core/lib/Thelia/Core/Template/ParserContext.php]: Indeed !

**Comment by roadster31** [comment]
In the back-office, most error handling is performed by redisplaying the current template, without redirecting, thus the form remains in the current context, and a session storage is not required.

However, many modules do not use this system, and if you give a look, you'll see that most of them, simply don't process form errors, because there's no way to do it !

The situation is the same in the front office.

Thus, IMO, this PR is not an "enhancement", but just a "temporary workaround". Additionnaly, we *should not* introduce parameters such as $pushInSeesion, which are only a way to partially compensate a major form lifecycle design flaw.

We should invest some time to work on form error management.

**Pull request closed without merging**